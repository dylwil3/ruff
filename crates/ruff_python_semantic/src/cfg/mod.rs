//! # Overview
//! To build a control-flow graph, we step through each statement in order.
//! By default, statements are added to the current basic block until we
//! reach a statement that invokes control flow. These are exactly the following:
//!
//! | Branching                  | Loops   | Jumps      |
//! |----------------------------|---------|------------|
//! | `if`                       | `for`   | `break`    |
//! | `match`                    | `while` | `continue` |
//! | `try`                      |         | `raise`    |
//! | `with`                     |         | `return`   |
//!
//! There are also `assert` statements, which are equivalent
//! to a branch followed by a jump. That is,
//! ```
//! assert cond
//! ```
//!
//! is equivalent to
//!
//! ```
//! if not cond:
//!   raise AssertionError
//! ```
//!
//! (Technically there is an additional `if __debug__` wrapped around, but
//! we will ignore that.)
//!
//! The control flow graph is then constructed using this equivalent
//! form.
//!
//!
//! We now discuss how each kind of control flow is handled.
//!
//! ## Branching
//!
//! Upon reaching a branching statement, the statement
//! itself terminates the basic block, and several outgoing
//! edges are added, labeled by the condition that needs to
//! be satisfied in order to traverse that edge. For example:
//!
//! ```python
//! if cond:
//!   f()
//! else:
//!   g()
//! ```
//! Produces:
//! ```text
//!          +--------+
//!          |  Start |
//!          +--------+
//!              |
//!              v
//!           +-------+
//!   +------>|if stmt|
//!   |       +-------+
//!   |          |
//!   | cond     | not cond
//!   |          |
//!   v          v
//! +----+     +----+
//! | f()|     | g()|
//! +----+     +----+
//!   |          |
//!   +-----+----+
//!         |
//!       +--------+
//!       |  End   |
//!       +--------+
//!```
//!
//! ## Loops
//!
//! A loop consists of a _loop guard_, a _body_, and an optional
//! _else_ clause. The _loop guard_ is a condition we check to
//! determine whether to re-enter the loop body, the _body_ is again
//! a list of statements, and the _else_ clause is a possible
//! exit from the loop body or guard (depending). For simplicity,
//! let's ignore the else clause for a moment and specialize to the
//! case of `while` loops.
//!
//! When we reach a `while` loop, we begin by creating a new
//! basic block that is comprised entirely of the `loop` guard,
//! and which unconditionally follows the preceding basic block.
//! The loop guard consists of the test clause for the `while` statement.
//! This loop guard will have an outgoing node to the _loop exit_
//! and another to the loop body; the edge followed corresponds to the
//! veracity of the test clause.
//!
//! The loop body will almost always have two outgoing edges: one
//! which points back to the loop guard, and another that goes to the
//! loop exit. An exception would be the case of a jump statement, like
//! `continue` or `raise`.
//!
//! For example,
//!
//! ```python
//! while cond:
//!   continue
//! ```
//! would create
//!
//!```text
//!          ┌─────────┐  
//!          │         │  
//!          │  Start  │  
//!          │         │  
//!          └────┬────┘  
//!               │       
//!          ┌────▼────┐  
//!          │         │  
//!     ┌───►│  cond?  │  
//!     │    │         │  
//!     │    └─────┬──┬┘  
//!     │          │  │   
//!     │          │  │   
//! ┌───┴───────┐  │  │   
//! │           │  │  │   
//! │ continue  │◄─┘  │   
//! │           │     │   
//! └───────────┘     │   
//!                   │   
//!                   │   
//!           ┌───────▼──┐
//!           │          │
//!           │Loop Exit │
//!           │          │
//!           └──────────┘
//! ```
//!
//! ## Jumps
//!
//! Upon reaching a jump statement, we terminate the basic block
//! and add an outgoing edge. The target of the outgoing edge is
//! determined by the _loop context_ and the _try context_. That is,
//! we need to know whether and where we are within a try-statement,
//! and what (if any) is the innermost loop surrounding the current
//! statement.
//!
//! ### Break and Continue
//!
//! When we encounter a `break` or `continue`, we first check the
//! _try context_ to see if we are in a try context that has a
//! `finally` clause. In this case, we resolve the `finally` clause
//! first. Assuming no further jumps have occurred, we then check the
//! current loop context and direct flow to the loop exit (resp. guard)
//! in the case of a `break` (resp. `continue`).
//!
//! ### Raise and Return
//!
//! When we encounter a `raise` or `return`, we first check the
//! _try context_ stack to see if we are in a try context, and
//! whether it has an `except` or `finally` clause. In the case of
//! a `return`, we first resolve the `finally` block before returning.
//! In the case of a `raise`, we visit `except` blocks and then the
//! `finally` block. Without type inference, we are limited in our
//! ability to determine which `except` block is visited. As a first
//! pass, we pretend that we visit each of them.
//!
//! # Implementation Details
//!
//! Our implementation is non-recursive, guaranteed to terminate,
//! and takes a single, forward pass through a list of statements.
//!
//! To achieve this, our builder maintains two pieces of information:
//! - A stack of loop contexts
//! - A stack of `try` contexts
//!
//! # Misc other things
//! - When we encounter raise/return we should make remaining
//! things in local block an orphan node of some kind and mark
//! unreachable? Or add to worklist for later or something?
//! I guess there could be some context about being unreachable
//! or something...? Or maybe no special casing is required? Unclear.
pub mod builder;
pub mod implementations;
pub mod visualize;
