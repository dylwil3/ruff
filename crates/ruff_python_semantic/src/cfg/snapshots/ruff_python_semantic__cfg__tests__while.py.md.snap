---
source: crates/ruff_python_semantic/src/cfg/mod.rs
description: "This is a Mermaid graph. You can use https://mermaid.live to visualize it as a diagram."
---
## Function 0
### Source
```python
def func():
    while False:
        return "unreachable"
    return 1
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return 1"]
	node3["EMPTY"]
	node4["return #quot;unreachable#quot;"]
	node0-->node3
	node2==>node1
	node3-->|"False"|node4
	node3-->|"Else"|node2
	node4==>node1
```

## Function 1
### Source
```python
def func():
    while False:
        return "unreachable"
    else:
        return 1
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["return #quot;unreachable#quot;"]
	node4["return 1"]
	node0-->node2
	node2-->|"False"|node3
	node2-->|"Else"|node4
	node3==>node1
	node4==>node1
```

## Function 2
### Source
```python
def func():
    while False:
        return "unreachable"
    else:
        return 1
    return "also unreachable"
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return #quot;also unreachable#quot;"]
	node3["EMPTY"]
	node4["return #quot;unreachable#quot;"]
	node5["return 1"]
	node0-->node3
	node2==>node1
	node3-->|"False"|node4
	node3-->|"Else"|node5
	node4==>node1
	node5==>node1
```

## Function 3
### Source
```python
def func():
    while True:
        return 1
    return "unreachable"
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return #quot;unreachable#quot;"]
	node3["EMPTY"]
	node4["return 1"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4==>node1
```

## Function 4
### Source
```python
def func():
    while True:
        return 1
    else:
        return "unreachable"
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["return 1"]
	node4["return #quot;unreachable#quot;"]
	node0-->node2
	node2-->|"True"|node3
	node2-->|"Else"|node4
	node3==>node1
	node4==>node1
```

## Function 5
### Source
```python
def func():
    while True:
        return 1
    else:
        return "unreachable"
    return "also unreachable"
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return #quot;also unreachable#quot;"]
	node3["EMPTY"]
	node4["return 1"]
	node5["return #quot;unreachable#quot;"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node5
	node4==>node1
	node5==>node1
```

## Function 6
### Source
```python
def func():
    i = 0
    while False:
        i += 1
    return i
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["i = 0"]
	node1((("EXIT")))
	node2["return i"]
	node3["EMPTY"]
	node4["i += 1"]
	node0-->node3
	node2==>node1
	node3-->|"False"|node4
	node3-->|"Else"|node2
	node4-->node3
```

## Function 7
### Source
```python
def func():
    i = 0
    while True:
        i += 1
    return i
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["i = 0"]
	node1((("EXIT")))
	node2["return i"]
	node3["EMPTY"]
	node4["i += 1"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->node3
```

## Function 8
### Source
```python
def func():
    while True:
        pass
    return 1
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return 1"]
	node3["EMPTY"]
	node4["pass"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->node3
```

## Function 9
### Source
```python
def func():
    i = 0
    while True:
        if True:
            print("ok")
        i += 1
    return i
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["i = 0"]
	node1((("EXIT")))
	node2["return i"]
	node3["EMPTY"]
	node4["EMPTY"]
	node5["i += 1"]
	node6["print(#quot;ok#quot;)"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->|"True"|node6
	node4-->|"Else"|node5
	node5-->node3
	node6-->node5
```

## Function 10
### Source
```python
def func():
    i = 0
    while True:
        if False:
            print("ok")
        i += 1
    return i
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["i = 0"]
	node1((("EXIT")))
	node2["return i"]
	node3["EMPTY"]
	node4["EMPTY"]
	node5["i += 1"]
	node6["print(#quot;ok#quot;)"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->|"False"|node6
	node4-->|"Else"|node5
	node5-->node3
	node6-->node5
```

## Function 11
### Source
```python
def func():
    while True:
        if True:
            return 1
    return 0
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["return 0"]
	node3["EMPTY"]
	node4["EMPTY"]
	node5["return 1"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->|"True"|node5
	node4-->|"Else"|node3
	node5==>node1
```

## Function 12
### Source
```python
def func():
    while True:
        continue
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["continue"]
	node0-->node2
	node2-->|"True"|node3
	node2==>|"Else"|node1
	node3-->node2
```

## Function 13
### Source
```python
def func():
    while False:
        continue
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["continue"]
	node0-->node2
	node2-->|"False"|node3
	node2==>|"Else"|node1
	node3-->node2
```

## Function 14
### Source
```python
def func():
    while True:
        break
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["break"]
	node0-->node2
	node2-->|"True"|node3
	node2==>|"Else"|node1
	node3==>node1
```

## Function 15
### Source
```python
def func():
    while False:
        break
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["break"]
	node0-->node2
	node2-->|"False"|node3
	node2==>|"Else"|node1
	node3==>node1
```

## Function 16
### Source
```python
def func():
    while True:
        if True:
            continue
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["EMPTY"]
	node4["continue"]
	node0-->node2
	node2-->|"True"|node3
	node2==>|"Else"|node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->node2
```

## Function 17
### Source
```python
def func():
    while True:
        if True:
            break
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["EMPTY"]
	node3["EMPTY"]
	node4["break"]
	node0-->node2
	node2-->|"True"|node3
	node2==>|"Else"|node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4==>node1
```

## Function 18
### Source
```python
def func():
    while True:
        x = 0
        x = 1
        break
        x = 2
    x = 3
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["x = 3"]
	node3["EMPTY"]
	node4["x = 0
x = 1
break"]
	node5["x = 2"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->node2
	node5-->node3
```

## Function 19
### Source
```python
def func():
    while True:
        x = 0
        x = 1
        continue
        x = 2
    x = 3
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["x = 3"]
	node3["EMPTY"]
	node4["x = 0
x = 1
continue"]
	node5["x = 2"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4-->node3
	node5-->node3
```

## Function 20
### Source
```python
def func():
    while True:
        x = 0
        x = 1
        return
        x = 2
    x = 3
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["EMPTY"]
	node1((("EXIT")))
	node2["x = 3"]
	node3["EMPTY"]
	node4["x = 0
x = 1
return"]
	node5["x = 2"]
	node0-->node3
	node2==>node1
	node3-->|"True"|node4
	node3-->|"Else"|node2
	node4==>node1
	node5-->node3
```
